src/test/fuzz/cluster_linearize.cpp

```diff
@@ -495,11 +494,93 @@ FUZZ_TARGET(clusterlin_ancestor_finder)
+    // confidence in SimpleCandidateFinder, so that it can be used to test SearchCandidateFinder
+    // below.
+
+    // Retrieve a depgraph from the fuzz input.
+    SpanReader reader(buffer);
+    DepGraph<TestBitSet> depgraph;
+    try {
+        reader >> Using<DepGraphFormatter>(depgraph);
+    } catch (const std::ios_base::failure&) {}
+
+    // Instantiate the SimpleCandidateFinder to be tested, and the ExhaustiveCandidateFinder it is
+    // being tested against.
+    SimpleCandidateFinder smp_finder(depgraph);
+    ExhaustiveCandidateFinder exh_finder(depgraph);
+
+    auto todo = TestBitSet::Fill(depgraph.TxCount());
+    while (todo.Any()) {
+        assert(!smp_finder.AllDone());
+        assert(!exh_finder.AllDone());
+
+        // Call SimpleCandidateFinder.
+        auto [found, iterations_done] = smp_finder.FindCandidateSet(MAX_SIMPLE_ITERATIONS);
+        bool optimal = (iterations_done != MAX_SIMPLE_ITERATIONS);
+
+        // Sanity check the result.
+        assert(iterations_done <= MAX_SIMPLE_ITERATIONS);
+        assert(found.transactions.Any());
+        assert(found.transactions.IsSubsetOf(todo));
+        assert(depgraph.FeeRate(found.transactions) == found.feerate);
+        // Check that it is topologically valid.
+        for (auto i : found.transactions) {
+            assert(found.transactions.IsSupersetOf(depgraph.Ancestors(i) & todo));
+        }
+
+        // At most 2^(N-1) iterations can be required: the number of non-empty connected subsets a
+        // graph with N transactions can have.
+        assert(iterations_done <= (uint64_t{1} << (todo.Count() - 1)));
+
+        // Perform quality checks only if SimpleCandidateFinder claims an optimal result.
+        if (optimal) {
+            // Optimal sets are always connected.
+            assert(depgraph.IsConnected(found.transactions));
+
+            if (todo.Count() <= 12) {
+                // Compare with ExhaustiveCandidateFinder. This quickly gets computationally
+                // expensive for large clusters (O(2^n)), so only do it for sufficiently small ones.
+                auto exhaustive = exh_finder.FindCandidateSet();
+                assert(exhaustive.feerate == found.feerate);
+            } else {
+                // For larger todo sets, just test against every ancestor set ...
+                for (auto t1 : todo) {
+                    const auto& a1 = depgraph.Ancestors(t1);
+                    assert(found.feerate >= depgraph.FeeRate(a1 & todo));
+                }
+                // and every union of two ancestor sets.
+                for (auto t1 : todo) {
+                    const auto& a1 = depgraph.Ancestors(t1);
+                    for (auto t2 : todo - a1) {
+                        const auto& a2 = depgraph.Ancestors(t2);
+                        assert(found.feerate >= depgraph.FeeRate((a1 | a2) & todo));
+                    }
```

Comment:
This can be in the same loop above?

Why not just compare with the result of ancestor finder?